### PR TITLE
LibC+Utilities: Add groupadd and groupdel

### PR DESCRIPTION
**Pull request by M4xi1m3**
This PR adds the `groupadd` and `groupdel` utilities. It also adds the `putgrent` LibC function. This closes issue #4898.

### CODE REVIEW

File: Userland/Utilities/groupadd.cpp [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        return 3;
+    }
+
+    // First, let's sort out the gid for the group
+    if (gid > 0) {
+        if (getgrgid(static_cast<uid_t>(gid))) {
+            warnln("gid {} already exists!", gid);
+            return 4;
+        }
+
```
Comment (boricj) [inline Userland/Utilities/groupadd.cpp:69]: Extraneous whitespace.

File: Userland/Utilities/groupdel.cpp [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        perror("unveil");
+        return 1;
+    }
+
+    if (unveil("/bin/rm", "x") < 0) {
+        perror("unveil");
+        return 1;
+    }
+
+    const char* groupname = nullptr;
```
Comment (boricj) [inline Userland/Utilities/groupdel.cpp:47]: Use east const style: `char const*`. This comment applies to the diff as a whole.

File: Userland/Utilities/groupdel.cpp [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    // Search if the group is the primary group of an user
+    setpwent();
+    struct passwd* pw;
+    for (pw = getpwent(); pw; pw = getpwent()) {
+        if (pw->pw_gid == gid)
+            break;
+    }
+
+    // If pw is not NULL it means we ended prematuraly, aka. the group was found as primary group of an user
+    if (pw != NULL) {
```
Comment (boricj) [inline Userland/Utilities/groupdel.cpp:74]: Use `nullptr` instead of `NULL`.

File: Userland/Libraries/LibC/grp.cpp [hunk 4]
```diff
@@ -159,4 +161,45 @@ int initgroups(const char* user, gid_t extra_gid)
         gids[count++] = extra_gid;
     return setgroups(count, gids);
 }
+
+int putgrent(const struct group* g, FILE* stream)
```
Comment (gunnarbeutner) [inline Userland/Libraries/LibC/grp.cpp:165]:
```suggestion
int putgrent(const struct group* group, FILE* stream)
```

I know `putpwent` also just calls this `p`, but that's not a good variable name. :)

File: Userland/Libraries/LibC/grp.h [hunk 5]
```diff
@@ -23,6 +24,7 @@ void setgrent();
 void endgrent();
 struct group* getgrnam(const char* name);
 struct group* getgrgid(gid_t);
+int putgrent(const struct group* p, FILE* stream);
```
Comment (gunnarbeutner) [inline Userland/Libraries/LibC/grp.h:27]: You can and should omit the variable name for arguments if their use is obvious, for example here. A non-obvious use would be a bool argument (unless the function is a setter).

File: Userland/Utilities/groupadd.cpp [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    args_parser.parse(argc, argv);
+
+    // Let's run a quick sanity check on groupname
+    if (strpbrk(groupname, "\\/!@#$%^&*()~+=`:\n")) {
+        warnln("invalid character in groupname, {}", groupname);
+        return 1;
+    }
+
+    // Disallow names starting with _ and -
+    if (groupname[0] == '_' || groupname[0] == '-' || !isalpha(groupname[0])) {
```
Comment (gunnarbeutner) [inline Userland/Utilities/groupadd.cpp:48]: Luckily `!isalpha(groupname[0])` is false for the NUL character (e.g. if the user specified an empty string. Should we make that more explicit though?

File: Userland/Utilities/groupdel.cpp [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+#include <pwd.h>
+#include <shadow.h>
+#include <spawn.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <unistd.h>
```
Comment (gunnarbeutner) [inline Userland/Utilities/groupdel.cpp:28]: Some of those includes are unnecessary, e.g.:

```
#include <dirent.h>
#include <shadow.h>
#include <spawn.h>
#include <sys/wait.h>
```

I haven't checked all of them though.

**Review by gunnarbeutner — CHANGES_REQUESTED** [review]

**Comment by M4xi1m3** [comment]
Thank you for your review. I added the changes you suggested and removed unnecessary includes.

File: Userland/Utilities/groupadd.cpp (hunk 6 above)
Comment (M4xi1m3) [reply Userland/Utilities/groupadd.cpp:41]: Do you mean documenting the `isalpha` function (out of the scope of that pr) or adding a comment saying we refuse empty string ?

Comment (gunnarbeutner) [reply Userland/Utilities/groupadd.cpp:41]: No, I mean, maybe check whether it's empty explicitly - seeing how the behavior of `isalpha` might not be obvious here.

**Pull request merged**